### ISSUE TITLE
[Bug]: PA_ND sparse_attn_sharedkv output depends on physical compressed-KV block allocation

### ISSUE THREAD

**Issue by QwertyJack** [vllm-project/vllm-ascend#14013]
### Your current environment

<details>
<summary>Sanitized reproducer environment</summary>

```text
Hardware: Ascend 910_93 (A3), TP8
vLLM Ascend: 50e0ce60849a7b6fe6f94e1de085b6fc06e7006d
vLLM: 752a3a504485790a2e8491cacbb35c137339ad34
CANN: 9.0.1, innerversion V100R001C10SPC002B220
torch: 2.10.0
torch_npu: 2.10.0.post2
Python: 3.12.13
Execution path: DeepSeek V4, non-PD, TP8, eager, EP disabled, MTP disabled, prefix caching disabled, DSA overlap disabled
```

The same long-context non-repeatability was also observed on Ascend 910B (A2) with BF16 weights. Customer-specific environment details and prompts are omitted.

</details>

## Summary

`sparse_attn_sharedkv` in the PA_ND compressed-KV path can produce different BF16 outputs when only the physical paged-KV block allocation changes. The logical query, K/V values, sparse indices, and attention metadata remain identical.

This makes DeepSeek V4 logits and selected tokens depend on allocator state. It reproduces without PD separation, graph mode, EP, MTP, prefix caching, or DSpark. Higher concurrency can exercise block reuse and fragmentation more often, but this issue does not make a quantitative failure-rate claim.

The operator defect and its fix are independently demonstrated below. Premature EOS and `stop_without_tool_call` are not used as correctness criteria for this issue because a separate frontend rendering defect in #14035 materially changes that boundary; see [Scope](#scope).

## Root cause

The first incorrect boundary is [`SASVectorBlock<SAST>::CopyInKv`](https://github.com/vllm-project/vllm-ascend/blob/50e0ce60849a7b6fe6f94e1de085b6fc06e7006d/csrc/attention/sparse_attn_sharedkv/op_kernel/arch32/sparse_attn_sharedkv_scfa_block_vector.h#L589-L643).

For a pair of compressed-KV blocks, the PA_ND fast path derives `srcStride` from the absolute difference between two physical GM offsets, starts from the lower physical address, and issues one strided DMA:

```cpp
int64_t blkTableSrcStride =
    Abs(keyOffset1 - keyOffset2) - sparseBlockSize * headDim;
int64_t startGmOffset = MinValid(keyOffset1, keyOffset2);
intriParams.blockCount = (keyOffset1 >= 0) + (keyOffset2 >= 0);
intriParams.srcStride = blkTableSrcStride * sizeof(KV_T);
DataCopyPad(kvMergUb_[...], cmpKvGm_[startGmOffset], intriParams, padParams);
```

The DMA therefore copies the lower physical address first instead of preserving the logical order supplied by `cmp_sparse_indices`. Paged allocation is runtime state and must not change the semantic operator result, but here it changes the merged K/V row order consumed by matmul and softmax.

The K/V values themselves are intact. A diagnostic kernel mirrored the merged BF16 rows after `CopyInKv`, and a CPU reconstruction followed the source code's exact addressing and fallback rules. For rank-4 query row 629, changing only the physical block table reversed 61 of 79 pair slots, or 122 row positions, while preserving the exact K/V row multiset. Both captured DMA buffers matched the CPU reconstruction bitwise.

The row permutation changes finite-precision reduction order. The first observed tensor difference is in `sparse_attn_sharedkv`, before it propagates through `wo_a`, `wo_b`, all-reduce, later layers, logits, and token selection.

BSND and TND do not use this paged physical-address ordering rule.

## Proposed fix

For PA_ND, copy the two blocks separately with `CopyInSingleKv` in the order supplied by `cmp_sparse_indices`. BSND and TND keep the existing path.

This is preferable to sorting each pair by logical sparse index because `cmp_sparse_indices` already defines the operator input order. A conditional implementation that retained the dual DMA when physical and logical order happened to agree was also allocation-invariant, but compiled slower than the simpler unconditional implementation.

### Operator correctness

A real layer-2 replay kept query, logical K/V values, sparse indices, and attention metadata fixed while independently changing CMP and SWA physical block allocation.

With the current implementation on rank 0, the baseline and fragmented allocations differed in 8 BF16 output elements with `max_abs=0.001953125`:

```text
baseline:   3888e5a489d930a9ece4ad28b12d56031661fea33478824930f08ce40379b835
fragmented: 9facb5f3b6013ce5ae6a7a4b96e00368dda5231a57e3964ff18d8834073bbbfd
```

With the proposed fix, baseline, CMP-fragmented, SWA-fragmented, and both-fragmented allocations were bitwise identical:

```text
rank 0: 9260cfaa913a4faadae6012fe90ef13504918bad3ef3b4ebac7e425f41bff6ba
rank 4: 41bdccbf4a270f20ec2b698cb0569b2d2eb31512e2780b122d3faf685ee6bc3b
```

The final source rebuild reproduced the validated operator binary hashes:

```text
4434e957f158b6b8135005894c8846de840e8e7c62d28d44a4f47438ea4d1ff9  SparseAttnSharedkv_7d53da980fa2b93856b96304fde51fcc.o
e37d73cc655db42b51bfb7caab4e360f1965a4e1441d9f262199147f1ea2ff5a  SparseAttnSharedkv_b4f65ead04fb4fc6e9a77e022317d6e5.o
```

### Service repeatability

The fixed build produced identical selected tokens and full top-20 logprobs across repeated token-in boundary probes:

| Case | Prompt tokens | Result |
| --- | ---: | --- |
| Frozen boundary | 57,578 | `8/8` identical EOS, chosen logprob, full top-20, and `finish_reason=stop` |
| Independent q022 boundary | 56,933 | `8/8` identical four-space token, chosen logprob, full top-20, and `finish_reason=length` |

The frozen-boundary EOS result used the pre-#14035 renderer and is repeatability evidence only. It is not evidence that EOS is the correct next token; see [Scope](#scope).

### Performance

Measurements used the same real layer-2 input on A3, with 10 warmups, 500 timed iterations, and device synchronization around each iteration.

| PA_ND implementation | rank-0 baseline | rank-0 fragmented | rank-4 baseline | rank-4 fragmented |
| --- | ---: | ---: | ---: | ---: |
| Current dual DMA | 3.033 ms | 3.034 ms | not measured | not measured |
| Proposed two-single-DMA fix | 3.207 ms | 3.208 ms | 3.208 ms | 3.213 ms |
| Conditional dual/single DMA | 3.347 ms | 3.350 ms | 3.367 ms | 3.368 ms |

The proposed fix is about 5.7% slower than the faulty implementation on rank 0, but 4.2%-4.7% faster than the conditional variant.

## Scope

This issue establishes a PA_ND operator-contract defect: physical compressed-KV allocation changes `sparse_attn_sharedkv` output. That defect remains valid regardless of the frontend renderer.

It does not claim that the SAS defect is the primary cause of premature EOS. The earlier EOS stress runs and frozen 57,578-token boundary used a renderer that placed tools and system content in the wrong order. With #14035, the rendered prompt matches the checkpoint reference at 56,894 tokens, and rebuilding the same reasoning boundary changes the local top-1 from EOS to the legal `;\n\n` continuation also selected by the official API. EOS remains a lower-ranked sampling candidate locally, so termination policy is being investigated separately.

The remaining contribution of the SAS defect to premature-EOS frequency will be measured with an old/new renderer x faulty/fixed SAS 2x2 experiment. That experiment is not required to establish the allocation-dependent operator bug reported here.

## Impact

The defect directly affects DeepSeek V4 SCFA when it uses PA_ND paged compressed KV on the `arch32` kernel. It was proven with an A3 real-input operator replay and observed through A3 service-level non-repeatability. The corresponding A2 long-context symptom was reproduced with BF16 weights on the same `arch32` operator path, but the sanitized operator-level replay was performed on A3.

The kernel also supports FP16 and graph execution, but those paths are not claimed as separately validated here.

## Affected versions

The faulty implementation was introduced in [`b4f0100986`](https://github.com/vllm-project/vllm-ascend/commit/b4f01009860b0b00e243e1eb95d0107b332c1bad) by #9228, enabled for Ascend 910B in [`d4b36710fe`](https://github.com/vllm-project/vllm-ascend/commit/d4b36710fef6cf14544b911d5af656378def55b9) by #9339, and used by the DeepSeek V4 runtime path added in [`7bce23cc05`](https://github.com/vllm-project/vllm-ascend/commit/7bce23cc0572a7793636ae215015a528e4503f1e) by #9270.

Affected releases start at `v0.20.2rc1` and include the current `releases/v0.25.1rc` branch.
